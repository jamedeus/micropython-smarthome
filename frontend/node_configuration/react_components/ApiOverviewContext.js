import React, { useState, createContext } from 'react';
import PropTypes from 'prop-types';


export const ApiOverviewContext = createContext();

export const ApiOverviewContextProvider = ({ children }) => {
    // Load context set by django template
    const [context, setContext] = useState(() => {
        function parse_dom_context(name) {
            const element = document.getElementById(name);
            if (element) {
                return JSON.parse(element.textContent);
            } else {
                return "";
            }
        }

        // Parse context elements created by django template
        return {
            nodes: parse_dom_context("nodes"),
            macros: parse_dom_context("macros"),
            recording: parse_dom_context("recording")
        };
    });

    // Create state for macro record mode, contains name of macro
    // being recorded (default loaded from django template context)
    const [recording, setRecording] = useState(context.recording);

    const deleteMacro = (name) => {
        let update = { ...context.macros };
        delete update[name];
        setContext({ ...context, ["macros"]: update});
    };

    const deleteMacroAction = (name, index) => {
        let update = { ...context.macros };
        delete update[name][index];
        // If last action deleted remove whole macro
        if (update[name].every(item => item === null)) {
            deleteMacro(name);
        // If actions remain update macro
        } else {
            setContext({ ...context, ["macros"]: update});
        }
    };

    return (
        <ApiOverviewContext.Provider value={{
            context,
            setContext,
            recording,
            setRecording,
            deleteMacro,
            deleteMacroAction
        }}>
            {children}
        </ApiOverviewContext.Provider>
    );
};

ApiOverviewContextProvider.propTypes = {
    children: PropTypes.node,
};
