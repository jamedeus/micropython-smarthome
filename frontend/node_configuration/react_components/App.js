import React from 'react';
import InstanceCard from './InstanceCard';
import { TransitionGroup, CSSTransition } from 'react-transition-group';
import './Animation.css';


// Takes object and key prefix, returns all keys that begin with prefix
function filterObject(obj, prefix) {
    return Object.entries(obj).reduce((acc, [key, value]) => {
        if (key.startsWith(prefix)) {
            acc[key] = value;
        }
        return acc;
    }, {});
};


class App extends React.Component {
    constructor(props) {
        super(props);
        // Default state object if not received from context
        this.state = {
            metadata: {},
            wifi: {}
        };
    }

    componentDidMount() {
        // Overwrite state with config received from context (if present)
        const config = JSON.parse(document.getElementById("config").textContent);
        console.log(config)
        if (config) {
            this.setState({ ...config });
        }
    }

    logState = () => {
        console.log(this.state);
    };

    // Handler for add device and add sensor buttons
    addInstance = (category) => {
        this.setState(prevState => {
            // Get index of new instance
            const index = Object.keys(filterObject(prevState, category)).length + 1;

            // Add key to state object with empty config
            // Will be populated with metadata template when user selects type
            prevState[`${category}${index}`] = {}

            return prevState;
        });
    };

    // Handler for delete button on device and sensor cards
    deleteInstance = (id) => {
        this.setState(prevState => {
            // Remove deleted instance from state object
            delete prevState[id]
            // Decrement all subsequent IDs in state object to prevent gaps in index
            return update_ids(id, prevState);
        });
    };

    // Handler for type select dropdown in device and sensor cards
    changeInstanceType = (id, category, event) => {
        this.setState(prevState => {
            // Get config template for selected type from metadata object
            const template = metadata[`${category}s`][event.target.value]['config_template'];
            // Overwrite section in state object with config template
            prevState[id] = template;
            return prevState;
        })
    };

    // Handler for all inputs inside device and sensor cards
    // Takes device/sensor ID, param, and value; updates state object
    // TODO this has issues for components with multiple inputs (on/off paths, thermostat)
    handleInputChange = (id, param, value) => {
        console.log(param)
        this.setState(prevState => {
            prevState[id][param] = value;
            return prevState;
        });
    };

    // Render 2 column layout with device and sensor cards
    renderLayout = () => {
        const deviceEntries = (
            <TransitionGroup className="device-container">
                {Object.entries(this.state)
                .filter(([key, _]) => key.startsWith('device'))
                .map(([key, config]) => (
                    <CSSTransition key={key} timeout={500} classNames="fade">
                        <InstanceCard
                            key={key}
                            id={key}
                            category="device"
                            config={config}
                            onDelete={() => this.deleteInstance(key)}
                            onInputChange={(paramName, value) => this.handleInputChange(key, paramName, value)}
                            onTypeChange={(event) => this.changeInstanceType(key, "device", event)}
                        />
                    </CSSTransition>
                ))}
            </TransitionGroup>
        );

        const sensorEntries = (
            <TransitionGroup className="sensor-container">
                {Object.entries(this.state)
                .filter(([key, _]) => key.startsWith('sensor'))
                .map(([key, config]) => (
                    <CSSTransition key={key} timeout={500} classNames="fade">
                        <InstanceCard
                            key={key}
                            id={key}
                            category="sensor"
                            config={config}
                            onDelete={() => this.deleteInstance(key)}
                            onInputChange={(paramName, value) => this.handleInputChange(key, paramName, value)}
                            onTypeChange={(event) => this.changeInstanceType(key, "sensor", event)}
                        />
                    </CSSTransition>
                ))}
            </TransitionGroup>
        );

        return (
            <>
                <h1 className="text-center pt-3 pb-4">TEST</h1>

                <div id="page1" className="d-flex flex-column h-100">
                    <div className="row mt-3">
                        <div id="sensors" className="col-sm">
                            <h2 className="text-center">Add Sensors</h2>
                            {sensorEntries}
                            <div id="add_sensor" className="text-center position-relative">
                                <button className="btn-secondary btn mb-3" onClick={() => this.addInstance('sensor')}>Add Sensor</button>
                            </div>
                        </div>
                        <div id="devices" className="col-sm">
                            <h2 className="text-center">Add Devices</h2>
                            {deviceEntries}
                            <div id="add_device" className="text-center position-relative">
                                <button className="btn-secondary btn mb-3" onClick={() => this.addInstance('device')}>Add Device</button>
                            </div>
                        </div>
                    </div>
                </div>

                <div className="bottom">
                    <button className="log-button" onClick={() => this.logState()}>Log State</button>
                </div>
            </>
        );
    }

    render() {
        return this.renderLayout();
    }
}


// Called by deleteInstance, decrements IDs of all subsequent instances to prevent gaps
// Example: If device2 is deleted, device3 becomes device2, device4 becomes device3, etc
function update_ids(target, state) {
    // Get category (device or sensor) and index of removed instance
    const category = target.replace(/[0-9]/g, '');
    const index = target.replace(/[a-zA-Z]/g, '');

    // Get list of all instances in same category
    var instances = filterObject(state, category);

    // If target is device get list of sensors (used to update target IDs)
    if (category === 'device') {
        var sensors = filterObject(state, 'sensor');
        // Remove device from all sensor targets
        for (const sensor in sensors) {
            state[sensor]['targets'] = state[sensor]['targets'].filter(item => item !== target);
        };
    };

    // Iterate all instances in category starting from the removed instance index
    for (let i=parseInt(index); i<Object.entries(instances).length+1; i++) {
        // Removed index now available, decrement next index by 1
        const new_id = `${category}${i}`;
        const old_id = `${category}${i+1}`;
        state[new_id] = JSON.parse(JSON.stringify(state[old_id]));
        delete state[old_id];

        // Decrement device index in sensor targets lists to match above
        if (category === 'device') {
            for (const sensor in sensors) {
                if (state[sensor]['targets'].includes(old_id)) {
                    state[sensor]['targets'] = state[sensor]['targets'].filter(item => item !== old_id);
                    state[sensor]['targets'].push(new_id);
                };
            };
        };
    }

    return state;
};


export default App;
