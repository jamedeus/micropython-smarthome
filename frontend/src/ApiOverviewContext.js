import React, { useState, createContext } from 'react';
import PropTypes from 'prop-types';
import { parse_dom_context } from 'util/django_util';

export const ApiOverviewContext = createContext();

export const ApiOverviewContextProvider = ({ children }) => {
    // Load context set by django template
    const [context, setContext] = useState(() => {
        return {
            nodes: parse_dom_context("nodes"),
            macros: parse_dom_context("macros"),
            recording: parse_dom_context("recording"),
            start_recording: false
        };
    });

    // Create state for macro record mode, contains name of macro
    // being recorded (default loaded from django template context)
    const [recording, setRecording] = useState(context.recording);

    // Create state to show loading overlay
    const [loading, setLoading] = useState(false);

    // Remove loading overlay when navigated to with browser back button
    window.onpageshow = function(event) {
        if (event.persisted) {
            setLoading(false);
        }
    };

    const deleteMacro = (name) => {
        let update = { ...context.macros };
        delete update[name];
        setContext({ ...context, ["macros"]: update});
    };

    const deleteMacroAction = (name, index) => {
        let update = { ...context.macros };
        delete update[name][index];
        // If last action deleted remove whole macro
        if (update[name].every(item => item === null)) {
            deleteMacro(name);
        // If actions remain update macro
        } else {
            setContext({ ...context, ["macros"]: update});
        }
    };

    // Takes new macro name, sets recording state, sets context param
    // that opens instruction modal if "don't show again" cookie not set
    const startRecording = (name) => {
        setRecording(name);
        setContext({ ...context, ["start_recording"]: true});
    };

    return (
        <ApiOverviewContext.Provider value={{
            context,
            setContext,
            recording,
            setRecording,
            loading,
            setLoading,
            deleteMacro,
            deleteMacroAction,
            startRecording
        }}>
            {children}
        </ApiOverviewContext.Provider>
    );
};

ApiOverviewContextProvider.propTypes = {
    children: PropTypes.node,
};
