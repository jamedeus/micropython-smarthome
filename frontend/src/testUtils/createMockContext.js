// Creates DOM elements containing JSON (mocks django json_script filter)
// Takes context id (arg passed to parse_dom_context) and json contents
const createMockContext = (id, json) => {
    const mockContext = document.createElement('div');
    mockContext.id = id;
    mockContext.textContent = JSON.stringify(json);
    document.body.appendChild(mockContext);
};

export default createMockContext;
