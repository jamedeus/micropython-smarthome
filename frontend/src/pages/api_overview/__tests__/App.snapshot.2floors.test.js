import React from 'react';
import App from '../ApiOverview';
import { ApiOverviewContextProvider } from 'root/ApiOverviewContext';
import createMockContext from 'src/testUtils/createMockContext';
import { mockNodes2floors, mockMacros } from './mockContext';

describe('App', () => {
    it('matches snapshot when nodes are on 2 floors', () => {
        // Create mock state objects
        createMockContext('nodes', mockNodes2floors);
        createMockContext('macros', mockMacros);
        createMockContext('recording', '');

        // Render App, confirm matches snapshot
        const component = render(
            <ApiOverviewContextProvider>
                <App />
            </ApiOverviewContextProvider>
        );
        expect(component).toMatchSnapshot();
    });
});
