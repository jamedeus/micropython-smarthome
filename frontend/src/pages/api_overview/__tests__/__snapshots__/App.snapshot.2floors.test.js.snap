// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot when nodes are on 2 floors 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="nodes"
    >
      {"1":["Bathroom","Kitchen","Living Room"],"2":["Thermostat","Bedroom"]}
    </div>
    <div
      id="macros"
    >
      {"late":[{"ip":"192.168.1.100","args":["set_rule","device1","519"],"node_name":"Bathroom","target_name":"Bathroom LEDs","action_name":"Set Rule 519"},{"ip":"192.168.1.100","args":["disable","device2",""],"node_name":"Bathroom","target_name":"Bathroom Lights","action_name":"Disable"},{"ip":"192.168.1.101","args":["set_rule","device1","665"],"node_name":"Kitchen","target_name":"Cabinet Lights","action_name":"Set Rule 665"},{"ip":"192.168.1.101","args":["disable","device2",""],"node_name":"Kitchen","target_name":"Overhead Lights","action_name":"Disable"},{"ip":"192.168.1.103","args":["set_rule","device1","42"],"node_name":"Bedroom","target_name":"Lights","action_name":"Set Rule 42"},{"ip":"192.168.1.103","args":["enable","sensor2",""],"node_name":"Bedroom","target_name":"Computer Sensor","action_name":"Enable"},{"ip":"192.168.1.102","args":["set_rule","device1","42"],"node_name":"Living Room","target_name":"Overhead Lights","action_name":"Set Rule 42"},{"ip":"192.168.1.102","args":["set_rule","device2","12"],"node_name":"Living Room","target_name":"Lamp","action_name":"Set Rule 12"},{"ip":"192.168.1.100","args":["disable","device3",""],"node_name":"Bathroom","target_name":"Entry Light","action_name":"Disable"},{"ip":"192.168.1.103","args":["set_rule","device3","193"],"node_name":"Bedroom","target_name":"Bias lights","action_name":"Set Rule 193"}],"bright":[{"ip":"192.168.1.103","args":["set_rule","device1","98"],"node_name":"Bedroom","target_name":"Lights","action_name":"Set Rule 98"},{"ip":"192.168.1.100","args":["disable","device1",""],"node_name":"Bathroom","target_name":"Bathroom LEDs","action_name":"Disable"},{"ip":"192.168.1.100","args":["enable","device2",""],"node_name":"Bathroom","target_name":"Bathroom Lights","action_name":"Enable"},{"ip":"192.168.1.100","args":["enable","device3",""],"node_name":"Bathroom","target_name":"Entry Light","action_name":"Enable"},{"ip":"192.168.1.101","args":["disable","device1",""],"node_name":"Kitchen","target_name":"Cabinet Lights","action_name":"Disable"},{"ip":"192.168.1.101","args":["enable","device2",""],"node_name":"Kitchen","target_name":"Overhead Lights","action_name":"Enable"},{"ip":"192.168.1.102","args":["set_rule","device1","100"],"node_name":"Living Room","target_name":"Overhead Lights","action_name":"Set Rule 100"},{"ip":"192.168.1.102","args":["set_rule","device2","100"],"node_name":"Living Room","target_name":"Lamp","action_name":"Set Rule 100"}]}
    </div>
    <div
      id="recording"
    >
      ""
    </div>
    <div>
      <div
        class="d-flex flex-column vh-100"
      >
        <div
          class="d-flex justify-content-between"
        >
          <button
            class="btn btn-primary"
            style="visibility: hidden;"
            type="button"
          >
            <i
              class="bi-list"
            />
          </button>
          <h1
            class="my-3 "
          >
            Api Overview
          </h1>
          <div
            class="my-auto dropdown"
          >
            <button
              aria-expanded="false"
              class="dropdown-toggle btn btn-light"
              id="settings-button"
              type="button"
            >
              <i
                class="bi-list"
              />
            </button>
          </div>
        </div>
        <div
          class="text-center section mt-3 mb-4 p-3"
        >
          <h5
            class="mb-3 fw-bold"
          >
            Upstairs
          </h5>
          <button
            class="m-1 btn btn-primary"
            type="button"
          >
            Thermostat
          </button>
          <button
            class="m-1 btn btn-primary"
            type="button"
          >
            Bedroom
          </button>
        </div>
        <div
          class="text-center section mt-3 mb-4 p-3"
        >
          <h5
            class="mb-3 fw-bold"
          >
            Downstairs
          </h5>
          <button
            class="m-1 btn btn-primary"
            type="button"
          >
            Bathroom
          </button>
          <button
            class="m-1 btn btn-primary"
            type="button"
          >
            Kitchen
          </button>
          <button
            class="m-1 btn btn-primary"
            type="button"
          >
            Living Room
          </button>
        </div>
        <h1
          class="text-center mt-5"
        >
          Macros
        </h1>
        <div
          class="text-center section p-3 mx-auto macro-container"
        >
          <div>
            <div
              class="d-flex mb-3"
              id="late"
            >
              <div
                class="macro-row btn-group"
                role="group"
              >
                <button
                  class="macro-button btn btn-primary"
                  type="button"
                >
                  <h3
                    class="macro-name"
                  >
                    Late
                  </h3>
                </button>
                <div
                  class="macro-options dropdown btn-group"
                  role="group"
                >
                  <button
                    aria-expanded="false"
                    class="dropdown-toggle btn btn-primary"
                    type="button"
                  >
                    <i
                      class="bi-gear-fill"
                    />
                  </button>
                </div>
              </div>
            </div>
            <div
              class="d-flex mb-3"
              id="bright"
            >
              <div
                class="macro-row btn-group"
                role="group"
              >
                <button
                  class="macro-button btn btn-primary"
                  type="button"
                >
                  <h3
                    class="macro-name"
                  >
                    Bright
                  </h3>
                </button>
                <div
                  class="macro-options dropdown btn-group"
                  role="group"
                >
                  <button
                    aria-expanded="false"
                    class="dropdown-toggle btn btn-primary"
                    type="button"
                  >
                    <i
                      class="bi-gear-fill"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            class="text-center mt-3"
          >
            <button
              class="mt-3 mx-auto btn btn-secondary"
              type="button"
            >
              <i
                class="bi-plus-lg"
              />
            </button>
          </div>
          <div
            class="p-3 collapse"
          >
            <div
              class="form-floating mb-3"
            >
              <div
                class="form-floating"
              >
                <input
                  class="form-control"
                  id="new-macro-name"
                  placeholder="New macro name"
                  type="text"
                  value=""
                />
                <div
                  class="invalid-feedback"
                >
                  Name already in use
                </div>
                <label>
                  New macro name
                </label>
              </div>
            </div>
            <button
              class="btn btn-success"
              disabled=""
              type="button"
            >
              Start Recording
            </button>
          </div>
        </div>
        <div
          class="d-flex mx-auto flex-column mt-auto py-4"
        >
          <button
            class="btn btn-secondary"
            type="button"
          >
            Manage
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="d-flex flex-column vh-100"
    >
      <div
        class="d-flex justify-content-between"
      >
        <button
          class="btn btn-primary"
          style="visibility: hidden;"
          type="button"
        >
          <i
            class="bi-list"
          />
        </button>
        <h1
          class="my-3 "
        >
          Api Overview
        </h1>
        <div
          class="my-auto dropdown"
        >
          <button
            aria-expanded="false"
            class="dropdown-toggle btn btn-light"
            id="settings-button"
            type="button"
          >
            <i
              class="bi-list"
            />
          </button>
        </div>
      </div>
      <div
        class="text-center section mt-3 mb-4 p-3"
      >
        <h5
          class="mb-3 fw-bold"
        >
          Upstairs
        </h5>
        <button
          class="m-1 btn btn-primary"
          type="button"
        >
          Thermostat
        </button>
        <button
          class="m-1 btn btn-primary"
          type="button"
        >
          Bedroom
        </button>
      </div>
      <div
        class="text-center section mt-3 mb-4 p-3"
      >
        <h5
          class="mb-3 fw-bold"
        >
          Downstairs
        </h5>
        <button
          class="m-1 btn btn-primary"
          type="button"
        >
          Bathroom
        </button>
        <button
          class="m-1 btn btn-primary"
          type="button"
        >
          Kitchen
        </button>
        <button
          class="m-1 btn btn-primary"
          type="button"
        >
          Living Room
        </button>
      </div>
      <h1
        class="text-center mt-5"
      >
        Macros
      </h1>
      <div
        class="text-center section p-3 mx-auto macro-container"
      >
        <div>
          <div
            class="d-flex mb-3"
            id="late"
          >
            <div
              class="macro-row btn-group"
              role="group"
            >
              <button
                class="macro-button btn btn-primary"
                type="button"
              >
                <h3
                  class="macro-name"
                >
                  Late
                </h3>
              </button>
              <div
                class="macro-options dropdown btn-group"
                role="group"
              >
                <button
                  aria-expanded="false"
                  class="dropdown-toggle btn btn-primary"
                  type="button"
                >
                  <i
                    class="bi-gear-fill"
                  />
                </button>
              </div>
            </div>
          </div>
          <div
            class="d-flex mb-3"
            id="bright"
          >
            <div
              class="macro-row btn-group"
              role="group"
            >
              <button
                class="macro-button btn btn-primary"
                type="button"
              >
                <h3
                  class="macro-name"
                >
                  Bright
                </h3>
              </button>
              <div
                class="macro-options dropdown btn-group"
                role="group"
              >
                <button
                  aria-expanded="false"
                  class="dropdown-toggle btn btn-primary"
                  type="button"
                >
                  <i
                    class="bi-gear-fill"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
        <div
          class="text-center mt-3"
        >
          <button
            class="mt-3 mx-auto btn btn-secondary"
            type="button"
          >
            <i
              class="bi-plus-lg"
            />
          </button>
        </div>
        <div
          class="p-3 collapse"
        >
          <div
            class="form-floating mb-3"
          >
            <div
              class="form-floating"
            >
              <input
                class="form-control"
                id="new-macro-name"
                placeholder="New macro name"
                type="text"
                value=""
              />
              <div
                class="invalid-feedback"
              >
                Name already in use
              </div>
              <label>
                New macro name
              </label>
            </div>
          </div>
          <button
            class="btn btn-success"
            disabled=""
            type="button"
          >
            Start Recording
          </button>
        </div>
      </div>
      <div
        class="d-flex mx-auto flex-column mt-auto py-4"
      >
        <button
          class="btn btn-secondary"
          type="button"
        >
          Manage
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
